client/jmx-adapter/src/main/java/org/jolokia/client/jmxadapter/JolokiaJmxConnector.java

```diff
@@ -0,0 +1,125 @@
+package org.jolokia.client.jmxadapter;
+
+import java.util.HashMap;
+import org.jolokia.client.J4pClientBuilder;
+
+import javax.management.ListenerNotFoundException;
+import javax.management.MBeanServerConnection;
+import javax.management.NotificationBroadcasterSupport;
+import javax.management.NotificationFilter;
+import javax.management.NotificationListener;
+import javax.management.remote.JMXConnectionNotification;
+import javax.management.remote.JMXConnector;
+import javax.management.remote.JMXServiceURL;
+import javax.security.auth.Subject;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.util.Collections;
+import java.util.Map;
+
+/**
+ * Handle MBeanServerConnection using the jolokia protocol
+ * IN PROGRESS: Figuring out how to best handle connection strings in a JMC setup
+ */
+public class JolokiaJmxConnector implements JMXConnector {
+  private final JMXServiceURL serviceUrl;
+  private final Map<String, ?> environment;
+  private RemoteJmxAdapter adapter;
+  private final NotificationBroadcasterSupport broadcasterSupport=new NotificationBroadcasterSupport();
+  private long clientNotifSeqNo=1L;
+  private String connectionId;
+
+  public JolokiaJmxConnector(JMXServiceURL serviceURL, Map<String, ?> environment) {
+    this.serviceUrl=serviceURL;
+    this.environment=environment;
+  }
+
+  @Override
+  public void connect() throws IOException {
+    connect(Collections.<String, Object>emptyMap());
+  }
+
+  private String prefixWithSlashIfNone(String urlPath) {
+    if(urlPath.startsWith("/")) {
+      return urlPath;
+    } else {
+      return "/" + urlPath;
+    }
+  }
+
+  @Override
+  @SuppressWarnings({"raw"})
+  public void connect(Map<String, ?> env) throws IOException {
+    if(!"jolokia".equals(this.serviceUrl.getProtocol())) {
+      throw new MalformedURLException("I only handle Jolokia service urls");
+    }
+    Map<String,Object> mergedEnv=new HashMap<String,Object>();
+    if(this.environment != null) {
+      mergedEnv.putAll(this.environment);
+    }
+    if(env != null) {
+      mergedEnv.putAll(env);
+    }
+    String internalProtocol="http";
+    if(String.valueOf(this.serviceUrl.getPort()).endsWith("443") || "true".equals(mergedEnv.get("jmx.remote.x.check.stub"))) {
```

Comment:
Might figure out something smarter here, if anyone has an idea. Guess we can also make use of the env passed in to allow clients to override, such as env.put("jolokia.protocol","http")

Reply:
And possibly configure timeouts etc.